File: TESTING/LIN/dchkab.f
      PROGRAM DCHKAB

*  -- LAPACK test routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

*  =====================================================================

      // .. Parameters ..
      int                NMAX;
      const              NMAX = 132 ;
      int                MAXIN;
      const              MAXIN = 12 ;
      int                MAXRHS;
      const              MAXRHS = 16 ;
      int                MATMAX;
      const              MATMAX = 30 ;
      int                NIN, NOUT;
      const              NIN = 5, NOUT = 6 ;
      int                LDAMAX;
      const              LDAMAX = NMAX ;
      // ..
      // .. Local Scalars ..
      bool               FATAL, TSTDRV, TSTERR;
      String             C1;
      String             C2;
      String             PATH;
      String             INTSTR;
      String             ALINE;
      int                I, IC, K, LDA, NM, NMATS, NNS, NRHS, NTYPES, VERS_MAJOR, VERS_MINOR, VERS_PATCH;
      double             EPS, S1, S2, THRESH;
      REAL               SEPS
      // ..
      // .. Local Arrays ..
      bool               DOTYPE( MATMAX );
      int                IWORK( NMAX ), MVAL( MAXIN ), NSVAL( MAXIN );
      double             A( LDAMAX*NMAX, 2 ), B( NMAX*MAXRHS, 2 ), RWORK( NMAX ), WORK( NMAX*MAXRHS*2 );
      REAL               SWORK(NMAX*(NMAX+MAXRHS))
      // ..
      // .. External Functions ..
      double             DLAMCH, DSECND;
      bool               LSAME, LSAMEN;
      REAL               SLAMCH
      // EXTERNAL LSAME, LSAMEN, DLAMCH, DSECND, SLAMCH
      // ..
      // .. External Subroutines ..
      // EXTERNAL ALAREQ, DDRVAB, DDRVAC, DERRAB, DERRAC, ILAVER
      // ..
      // .. Scalars in Common ..
      bool               LERR, OK;
      String             SRNAMT;
      int                INFOT, NUNIT;
      // ..
      // .. Common blocks ..
      COMMON             / INFOC / INFOT, NUNIT, OK, LERR
      COMMON             / SRNAMC / SRNAMT
      // ..
      // .. Data statements ..
      DATA               INTSTR / '0123456789' /
      // ..
      // .. Executable Statements ..

      S1 = DSECND( )
      LDA = NMAX
      FATAL = .FALSE.

      // Read a dummy line.

      READ( NIN, FMT = * )

      // Report values of parameters.

      ilaver(VERS_MAJOR, VERS_MINOR, VERS_PATCH );
      WRITE( NOUT, FMT = 9994 ) VERS_MAJOR, VERS_MINOR, VERS_PATCH

      // Read the values of M

      READ( NIN, FMT = * )NM
      if ( NM.LT.1 ) {
         WRITE( NOUT, FMT = 9996 )' NM ', NM, 1
         NM = 0
         FATAL = .TRUE.
      } else if ( NM.GT.MAXIN ) {
         WRITE( NOUT, FMT = 9995 )' NM ', NM, MAXIN
         NM = 0
         FATAL = .TRUE.
      }
      READ( NIN, FMT = * )( MVAL( I ), I = 1, NM )
      for (I = 1; I <= NM; I++) { // 10
         if ( MVAL( I ).LT.0 ) {
            WRITE( NOUT, FMT = 9996 )' M  ', MVAL( I ), 0
            FATAL = .TRUE.
         } else if ( MVAL( I ).GT.NMAX ) {
            WRITE( NOUT, FMT = 9995 )' M  ', MVAL( I ), NMAX
            FATAL = .TRUE.
         }
      } // 10
      IF( NM.GT.0 ) WRITE( NOUT, FMT = 9993 )'M   ', ( MVAL( I ), I = 1, NM )

      // Read the values of NRHS

      READ( NIN, FMT = * )NNS
      if ( NNS.LT.1 ) {
         WRITE( NOUT, FMT = 9996 )' NNS', NNS, 1
         NNS = 0
         FATAL = .TRUE.
      } else if ( NNS.GT.MAXIN ) {
         WRITE( NOUT, FMT = 9995 )' NNS', NNS, MAXIN
         NNS = 0
         FATAL = .TRUE.
      }
      READ( NIN, FMT = * )( NSVAL( I ), I = 1, NNS )
      for (I = 1; I <= NNS; I++) { // 30
         if ( NSVAL( I ).LT.0 ) {
            WRITE( NOUT, FMT = 9996 )'NRHS', NSVAL( I ), 0
            FATAL = .TRUE.
         } else if ( NSVAL( I ).GT.MAXRHS ) {
            WRITE( NOUT, FMT = 9995 )'NRHS', NSVAL( I ), MAXRHS
            FATAL = .TRUE.
         }
      } // 30
      IF( NNS.GT.0 ) WRITE( NOUT, FMT = 9993 )'NRHS', ( NSVAL( I ), I = 1, NNS )

      // Read the threshold value for the test ratios.

      READ( NIN, FMT = * )THRESH
      WRITE( NOUT, FMT = 9992 )THRESH

      // Read the flag that indicates whether to test the driver routine.

      READ( NIN, FMT = * )TSTDRV

      // Read the flag that indicates whether to test the error exits.

      READ( NIN, FMT = * )TSTERR

      if ( FATAL ) {
         WRITE( NOUT, FMT = 9999 )
         STOP
      }

      // Calculate and print the machine dependent constants.

      SEPS = SLAMCH( 'Underflow threshold' )
      WRITE( NOUT, FMT = 9991 )'(single precision) underflow', SEPS
      SEPS = SLAMCH( 'Overflow threshold' )
      WRITE( NOUT, FMT = 9991 )'(single precision) overflow ', SEPS
      SEPS = SLAMCH( 'Epsilon' )
      WRITE( NOUT, FMT = 9991 )'(single precision) precision', SEPS
      WRITE( NOUT, FMT = * )

      EPS = DLAMCH( 'Underflow threshold' )
      WRITE( NOUT, FMT = 9991 )'(double          ) underflow', EPS;
      EPS = DLAMCH( 'Overflow threshold' )
      WRITE( NOUT, FMT = 9991 )'(double          ) overflow ', EPS;
      EPS = DLAMCH( 'Epsilon' )
      WRITE( NOUT, FMT = 9991 )'(double          ) precision', EPS;
      WRITE( NOUT, FMT = * )

      } // 80

      // Read a test path and the number of matrix types to use.

      READ( NIN, FMT = '(A72)', END = 140 )ALINE
      PATH = ALINE( 1: 3 )
      NMATS = MATMAX
      I = 3
      } // 90
      I = I + 1
      if ( I.GT.72 ) {
         NMATS = MATMAX
         GO TO 130
      }
      IF( ALINE( I: I ).EQ.' ' ) GO TO 90
      NMATS = 0
      } // 100
      C1 = ALINE( I: I )
      for (K = 1; K <= 10; K++) { // 110
         if ( C1.EQ.INTSTR( K: K ) ) {
            IC = K - 1
            GO TO 120
         }
      } // 110
      GO TO 130
      } // 120
      NMATS = NMATS*10 + IC
      I = I + 1
      IF( I.GT.72 ) GO TO 130
      GO TO 100
      } // 130
      C1 = PATH( 1: 1 )
      C2 = PATH( 2: 3 )
      NRHS = NSVAL( 1 )

      // Check first character for correct precision.

      IF( .NOT.LSAME( C1, 'double          ' ) ) THEN;
         WRITE( NOUT, FMT = 9990 )PATH


      } else if ( NMATS.LE.0 ) {

         // Check for a positive number of tests requested.

         WRITE( NOUT, FMT = 9989 )PATH
         GO TO 140

      } else if ( LSAMEN( 2, C2, 'GE' ) ) {

         // GE:  general matrices

         NTYPES = 11
         alareq('DGE', NMATS, DOTYPE, NTYPES, NIN, NOUT );

         // Test the error exits

         IF( TSTERR ) CALL DERRAB( NOUT )

         if ( TSTDRV ) {
            ddrvab(DOTYPE, NM, MVAL, NNS, NSVAL, THRESH, LDA, A( 1, 1 ), A( 1, 2 ), B( 1, 1 ), B( 1, 2 ), WORK, RWORK, SWORK, IWORK, NOUT );
         } else {
            WRITE( NOUT, FMT = 9989 )'DSGESV'
         }

      } else if ( LSAMEN( 2, C2, 'PO' ) ) {

         // PO:  positive definite matrices

         NTYPES = 9
         alareq('DPO', NMATS, DOTYPE, NTYPES, NIN, NOUT );


         IF( TSTERR ) CALL DERRAC( NOUT )


         if ( TSTDRV ) {
            ddrvac(DOTYPE, NM, MVAL, NNS, NSVAL, THRESH, LDA, A( 1, 1 ), A( 1, 2 ), B( 1, 1 ), B( 1, 2 ), WORK, RWORK, SWORK, NOUT );
         } else {
            WRITE( NOUT, FMT = 9989 )PATH
         }
      } else {

      }

      // Go back to get another input line.

      GO TO 80

      // Branch to this line when the last record is read.

      } // 140
      CLOSE ( NIN )
      S2 = DSECND( )
      WRITE( NOUT, FMT = 9998 )
      WRITE( NOUT, FMT = 9997 )S2 - S1

 9999 FORMAT( / ' Execution not attempted due to input errors' )
 9998 FORMAT( / ' End of tests' )
 9997 FORMAT( ' Total time used = ', F12.2, ' seconds', / )
 9996 FORMAT( ' Invalid input value: ', A4, '=', I6, '; must be >=', I6 )
 9995 FORMAT( ' Invalid input value: ', A4, '=', I6, '; must be <=', I6 )
 9994 FORMAT( ' Tests of the double           LAPACK DSGESV/DSPOSV',; ' routines ', / ' LAPACK VERSION ', I1, '.', I1, '.', I1, / / ' The following parameter values will be used:' )
 9993 FORMAT( 4X, A4, ':  ', 10I6, / 11X, 10I6 )
 9992 FORMAT( / ' Routines pass computational tests if test ratio is ', 'less than', F8.2, / )
 9991 FORMAT( ' Relative machine ', A, ' is taken to be', D16.6 )
 9990 FORMAT( / 1X, A6, ' routines were not tested' )
 9989 FORMAT( / 1X, A6, ' driver routines were not tested' )

      // End of DCHKAB

      }

File: test/lin/dchkab.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'dart:io';
import 'dart:math';

import 'package:dart_lapack/src/box.dart';
import 'package:dart_lapack/src/format_specifiers_extensions.dart';
import 'package:dart_lapack/src/install/dlamch.dart';
import 'package:dart_lapack/src/install/dsecnd.dart';
import 'package:dart_lapack/src/install/ilaver.dart';
import 'package:dart_lapack/src/install/lsame.dart';
import 'package:dart_lapack/src/install/slamch.dart';
import 'package:dart_lapack/src/lsamen.dart';
import 'package:dart_lapack/src/matrix.dart';
import 'package:dart_lapack/src/nio.dart';

import 'alareq.dart';
import 'ddrvab.dart';
import 'ddrvac.dart';
import 'derrab.dart';
import 'derrac.dart';

void main() async {
  final NIN = Nin(stdin), NOUT = Nout(stdout);
  const NMAX = 132;
  const MAXIN = 12;
  const MAXRHS = 16;
  const MATMAX = 30;
  const LDAMAX = NMAX;
  final DOTYPE = Array<bool>(MATMAX);
  final IWORK = Array<int>(NMAX),
      MVAL = Array<int>(MAXIN),
      NSVAL = Array<int>(MAXIN);
  final A = Matrix<double>(LDAMAX * NMAX, 2),
      B = Matrix<double>(NMAX * MAXRHS, 2),
      RWORK = Array<double>(NMAX),
      WORK = Array<double>(NMAX * MAXRHS * 2),
      SWORK = Array<double>(NMAX * (NMAX + MAXRHS));
  const INTSTR = '0123456789';

  final S1 = dsecnd();
  final LDA = NMAX;
  var FATAL = false;

  // Read a dummy line.

  await NIN.readLine();

  // Report values of parameters.

  final VERS_MAJOR = Box(0), VERS_MINOR = Box(0), VERS_PATCH = Box(0);
  ilaver(VERS_MAJOR, VERS_MINOR, VERS_PATCH);
  NOUT.println(
      ' Tests of the DOUBLE PRECISION LAPACK DSGESV/DSPOSV routines \n LAPACK VERSION ${VERS_MAJOR.value.i1}.${VERS_MINOR.value.i1}.${VERS_PATCH.value.i1}\n\n The following parameter values will be used:');

  // Read the values of M

  var NM = await NIN.readInt();
  if (NM < 1) {
    NOUT.print9996(' NM ', NM, 1);
    NM = 0;
    FATAL = true;
  } else if (NM > MAXIN) {
    NOUT.print9995(' NM ', NM, MAXIN);
    NM = 0;
    FATAL = true;
  }
  await NIN.readArray(MVAL, NM);
  for (var I = 1; I <= NM; I++) {
    if (MVAL[I] < 0) {
      NOUT.print9996(' M  ', MVAL[I], 0);
      FATAL = true;
    } else if (MVAL[I] > NMAX) {
      NOUT.print9995(' M  ', MVAL[I], NMAX);
      FATAL = true;
    }
  }
  if (NM > 0) NOUT.print9993('M   ', MVAL, NM);

  // Read the values of NRHS

  var NNS = await NIN.readInt();
  if (NNS < 1) {
    NOUT.print9996(' NNS', NNS, 1);
    NNS = 0;
    FATAL = true;
  } else if (NNS > MAXIN) {
    NOUT.print9995(' NNS', NNS, MAXIN);
    NNS = 0;
    FATAL = true;
  }
  await NIN.readArray(NSVAL, NNS);
  for (var I = 1; I <= NNS; I++) {
    if (NSVAL[I] < 0) {
      NOUT.print9996('NRHS', NSVAL[I], 0);
      FATAL = true;
    } else if (NSVAL[I] > MAXRHS) {
      NOUT.print9995('NRHS', NSVAL[I], MAXRHS);
      FATAL = true;
    }
  }
  if (NNS > 0) NOUT.print9993('NRHS', NSVAL, NNS);

  // Read the threshold value for the test ratios.

  final THRESH = await NIN.readDouble();
  NOUT.println(
      '\n Routines pass computational tests if test ratio is less than${THRESH.f8_2}\n');

  // Read the flag that indicates whether to test the driver routine.

  final TSTDRV = await NIN.readBool();

  // Read the flag that indicates whether to test the error exits.

  final TSTERR = await NIN.readBool();

  if (FATAL) {
    NOUT.println('\n Execution not attempted due to input errors');
    return;
  }

  // Calculate and print the machine dependent constants.

  var SEPS = slamch('Underflow threshold');
  NOUT.print9991('(single precision) underflow', SEPS);
  SEPS = slamch('Overflow threshold');
  NOUT.print9991('(single precision) overflow ', SEPS);
  SEPS = slamch('Epsilon');
  NOUT.print9991('(single precision) precision', SEPS);
  NOUT.println();

  var EPS = dlamch('Underflow threshold');
  NOUT.print9991('(double precision) underflow', EPS);
  EPS = dlamch('Overflow threshold');
  NOUT.print9991('(double precision) overflow ', EPS);
  EPS = dlamch('Epsilon');
  NOUT.print9991('(double precision) precision', EPS);
  NOUT.println();

  while (true) {
    // Read a test path and the number of matrix types to use.
    final String ALINE;
    try {
      ALINE = await NIN.readLine();
    } on EOF catch (_) {
      break;
    }
    final PATH = ALINE.substring(0, 3);
    var NMATS = MATMAX;
    var I = 3;
    do {
      I++;
      if (I > 72) {
        NMATS = MATMAX;
        break;
      }
    } while (ALINE[I - 1] == ' ');

    if (I <= 72) {
      NMATS = 0;
      while (true) {
        final C1 = ALINE[I - 1];
        var isDigit = false;
        var IC = 0;
        for (var K = 1; K <= 10; K++) {
          if (C1 == INTSTR[K - 1]) {
            IC = K - 1;
            isDigit = true;
            break;
          }
        }
        if (!isDigit) break;

        NMATS = NMATS * 10 + IC;
        I++;
        if (I > 72) break;
      }
    }
    final C1 = PATH[0];
    final C2 = PATH.substring(1, 3);
    // final NRHS = NSVAL[1];

    // Check first character for correct precision.

    if (!lsame(C1, 'Double precision')) {
      NOUT.println('\n ${PATH.a6} routines were not tested');
    } else if (NMATS <= 0) {
      // Check for a positive number of tests requested.

      NOUT.print9989(PATH);
      break;
    } else if (lsamen(2, C2, 'GE')) {
      // GE:  general matrices

      final NTYPES = 11;
      await alareq('DGE', NMATS, DOTYPE, NTYPES, NIN, NOUT);

      // Test the error exits

      if (TSTERR) derrab(NOUT);

      if (TSTDRV) {
        ddrvab(
            DOTYPE,
            NM,
            MVAL,
            NNS,
            NSVAL,
            THRESH,
            LDA,
            A(1, 1).asArray(),
            A(1, 2).asArray(),
            B(1, 1).asArray(),
            B(1, 2).asArray(),
            WORK,
            RWORK,
            SWORK,
            IWORK,
            NOUT);
      } else {
        NOUT.print9989('DSGESV');
      }
    } else if (lsamen(2, C2, 'PO')) {
      // PO:  positive definite matrices

      final NTYPES = 9;
      await alareq('DPO', NMATS, DOTYPE, NTYPES, NIN, NOUT);

      if (TSTERR) derrac(NOUT);

      if (TSTDRV) {
        ddrvac(
            DOTYPE,
            NM,
            MVAL,
            NNS,
            NSVAL,
            THRESH,
            LDA,
            A(1, 1).asArray(),
            A(1, 2).asArray(),
            B(1, 1).asArray(),
            B(1, 2).asArray(),
            WORK,
            RWORK,
            SWORK,
            NOUT);
      } else {
        NOUT.print9989(PATH);
      }
    } else {}

    // Go back to get another input line.
  }

  // Branch to this line when the last record is read.

  await NIN.close();
  final S2 = dsecnd();
  NOUT.println('\n End of tests');
  NOUT.println(' Total time used = ${(S2 - S1).f12_2} seconds\n');
}

extension on Nout {
  void print9996(String s, int actual, int expected) {
    println(
        ' Invalid input value: ${s.a4}=${actual.i6}; must be >=${expected.i6}');
  }

  void print9995(String s, int actual, int expected) {
    println(
        ' Invalid input value: ${s.a4}=${actual.i6}; must be <=${expected.i6}');
  }

  void print9993(final String s, final Array<int> a, int n) {
    var prefix = '    $s:  ';
    var i = 1;
    while (n > 0) {
      println('$prefix${a(i).i6(min(n, 10))}');
      prefix = ' ' * 11;
      n -= 10;
      i += 10;
    }
  }

  void print9991(String s, double EPS) {
    println(' Relative machine $s is taken to be${EPS.d16_6}');
  }

  void print9989(String s) {
    println('\n ${s.a6} driver routines were not tested');
  }
}
